Functions/Quantification/xASL_quant_HadamardDecoding.m

```diff
@@ -107,98 +125,141 @@
+decodedBlock = zeros(size(imEncodedReordered,1), size(imEncodedReordered,2), size(imEncodedReordered,3), xQ.TimeEncodedMatrixSize-1); % Temporary single decoded Hadamard block
 
-%% 4. Reorder multi-TE back to the initial order of PLD/TE
-%
-% For model fitting, we want the PLDs-first-TEs-second order (just like the
-% beginning) so we need to reorder it again
+% Go through all repetitions of the Hadamard blocks
+for iRepetition = 1:(nEncodedVolumes / xQ.TimeEncodedMatrixSize)
+	offsetEncoded = xQ.TimeEncodedMatrixSize * (iRepetition-1); % Index offset for the given repetition to search in the encoded data
+	encodedBlock = imEncoded(:,:,:,(1:xQ.TimeEncodedMatrixSize) + offsetEncoded); % Obtain the single encoded block
+	
+	% Decoded the single block
+	for iTI = 1:(xQ.TimeEncodedMatrixSize-1) %Number of decoded PLDs is number of Hadamard matrix size - 1
+		indexPositive = find(xQ.TimeEncodedMatrix(iTI,:)==1);
+		indexNegative = find(xQ.TimeEncodedMatrix(iTI,:)==-1);
+		decodedBlock(:,:,:,iTI) = sum(encodedBlock(:,:,:,(indexPositive)),4) - sum(encodedBlock(:,:,:,(indexNegative)),4);
+	end
+	offsetDecoded = (iRepetition-1)*(xQ.TimeEncodedMatrixSize-1); % Index offset for the given repetition in the decoded data
+	decodedPWI4D(:,:,:,offsetDecoded + (1:(xQ.TimeEncodedMatrixSize-1))) = decodedBlock; % Save the decoded block
+end
+
+%% 6. Reorder multi-TE back to the initial order of PLD/TE
+% For model fitting, we want the PLDs-first-TEs-second order (just like at the beginning) so we need to reorder it again
 if isfield(xQ,'NumberEchoTimes') && (xQ.NumberEchoTimes > 1)
 	% This is the original order
-	vectorOldOrder = 1:size(imDecoded, 4);
+	vectorReorderDecoded = 1:size(decodedPWI4D, 4);
 	
 	% Shape to a matrix with TEs second and all the rest first
-	vectorOldOrder = reshape(vectorOldOrder, size(imDecoded, 4)/xQ.NumberEchoTimes, xQ.NumberEchoTimes);
+	vectorReorderDecoded = reshape(vectorReorderDecoded, size(decodedPWI4D, 4)/xQ.NumberEchoTimes, xQ.NumberEchoTimes);
 	
 	% Flip the two dimensions and make a row vector again
-	vectorOldOrder = reshape(vectorOldOrder', 1, size(imDecoded, 4));
+	vectorReorderDecoded = reshape(vectorReorderDecoded', 1, size(decodedPWI4D, 4));
 
 	% Reorder the data
-	imDecoded = imDecoded(:,:,:,vectorOldOrder);
+	decodedPWI4D = decodedPWI4D(:,:,:,vectorReorderDecoded);
 end
 
-%% 5. Normalization of the decoded data
+% Calculate the correct parameters
+indexDecoding = ((xQ.NumberEchoTimes+1):nVolumesPerRepetition)'*ones(1,nRepetitions) + ones(nVolumesPerRepetition-xQ.NumberEchoTimes,1)*(0:nRepetitions-1)*nVolumesPerRepetition;
+indexDecoding = indexDecoding(:)';
+xQ.EchoTime_PWI4D = xQ.EchoTime(indexDecoding);
+xQ.InitialPLD_PWI4D = xQ.Initial_PLD(indexDecoding);
+xQ.LabelingDuration_PWI4D = xQ.LabelingDuration(indexDecoding);
+
+%% 7. Normalization of the decoded data
```

Comment:
Doesn't it make more sense to move this normalization to section 5 where you decode Hadamard?

Reply:
Yes, that is a good idea. I will do that.